design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,162 @@
+   1. Mark uncommitted data
+      1. List branches
+      2. Run _PrepareUncommittedForGC_ on all branches
+   2. Get all uncommitted data addresses
+      1. For each branch
+         1. Read all addresses from branch `BranchUncommittedMetarange` -> `Uncommitted DF`
+    >**Note:** To avoid possible bug, uncommitted listing must happen before listing of committed data
+3. Listing of lakeFS repository committed objects
+   1. Get all committed data addresses
+      1. Read all addresses from Repository commits -> `Committed DF`
+4. Find candidates for deletion
+   1. Subtract committed data from all objects (`Store DF` - `Committed DF`)
+   2. Subtract uncommitted data from all objects (`Store DF` - `Uncommitted DF`)
+   3. Filter files in special paths
+   4. The remainder is a list of files which can be safely removed
+5. Save run information under the GC path
+    1. The current run's `Uncommitted DF` (as a parquet file)
+    2. The last read partition
+    3. The GC run start timestamp
+
+#### Flow 2: Optimized Run
+
+Optimized run uses the previous GC run output, to perform a partial scan of the branch to remove uncommitted garbage.
+
+##### Step 1. Analyze Data and Perform Cleanup for old entries (GC client)
+
+1. Read previous run's information
+   1. Previous `Uncommitted DF`
+   2. Last read partition
+   3. Last run's timestamp
+2. Listing of lakeFS repository uncommitted objects  
+    See previous for steps
+3. Listing of lakeFS repository committed objects (optimized)
+   1. Read addresses from branch's new commits (all new commits down to the last GC run timestamp) -> `Committed DF`
+4. Find candidates for deletion
+   1. Subtract `Committed DF` from previous run's `Uncommitted DF`
+   2. Subtract current `Uncommitted DF` from previous run's `Uncommitted DF`
+   3. The result is a list of files that can be safely removed
+
+>**Note:** This step handles cases of objects that were uncommitted during previous GC run and are now deleted
+
+##### Step 2. Analyze Data and Perform Cleanup for new entries (GC client)
+
+1. Listing namespace objects (optimized)
+   1. Read all objects directly from object store
+   2. Skip partitions that are newer than < TOKEN_EXPIRY_TIME >
+   3. Using the partitions, stop after reading the last partition read by previous GC run -> `Store DF`
+2. Find candidates for deletion
+   1. Subtract `Committed DF` from `Store DF`
+   2. Subtract current `Uncommitted DF` from `Store DF`
+   3. Filter files in special paths
+   4. The remainder is a list of files which can be safely removed
+3. Save run information under the GC path  
+   See previous for steps
+
+## Limitations
+
+* Since this solution relies on the new repository structure, it is not backwards compatible. Therefore, another solution will be required for existing 
+repositories
+* For GC to work optimally, it must be executed on a timely basis. Long periods between runs might result in excessively long runs.
```

Comment:
Can we add some minimal required performance?  It will help both our implementation and users' sizing.